### PR TITLE
Support multiple parameter groups in step function, aligning with AdamW

### PR DESCRIPTION
**Pull request by sangoi-exe**
Hello! This is a suggested implementation to support multiple parameter groups like AdamW.

I modified the step function to iterate individually over each parameter group, instead of using only the first one for all the statistical calculations and so on.

This change allows it to handle multiple parameter groups with different settings (aka LR) seamlessly.

Why is this important? With Prodigy being widely used by the diffusion model community and becoming more of a standard, in addition to allowing for much more granular adjustment of hyperparameters, will correct the discrepancy that occurs when more than one neural network is trained together.

Currently, if the Unet (or its variant in more recent models) is trained alongside with one or more Text Encoders, the loss function calculation mixes the gradients and predictions all together. This leads to undertraining of the Unet and overtraining of the TE, as their weight adjustment requirements are on completely different scales.

If my modification is accepted, Prodigy will be able to train both the Unet and the TE flawlessly, as well as any other combination of neural networks, and the user won’t even realize anything has changed, aside from the improved training quality, as no default settings will be changed. 😉

Btw, I've been using this modification for days, and it works perfectly in practice, with no apparent increase in vram usage.

### CODE REVIEW

**Comment by adefazio** [comment]
Hi! Thanks for submitting this. Konstantin and I will take a look and get back to you on it.

**Comment by adefazio** [comment]
Sorry for the slow response here we are working on finalizing our Neurips submissions. Should have a chance to test this in the next few weeks.

**Comment by sangoi-exe** [comment]
> Sorry for the slow response here we are working on finalizing our Neurips submissions. Should have a chance to test this in the next few weeks.

Sure, dont worry, take your time

**Comment by dxqbYD** [comment]
This would be a useful addition!
To add some values:
common LRs used for training the SDXL unet are around 1e-05. Prodigy successfully finds a similar learning rate when used on unet only.

Training the CLIP text encoder requires LRs at about 1 magnitude less, I've seen between 1e-06 and 3e-06 being used.
I haven't tried this myself, but I would also expect that training fails using Prodigy on both networks as @sangoi-exe describes

**Comment by Arcitec** [comment]
Looks absolutely GREAT! What a great idea!

For anyone curious, this is a more generalized replacement for a previous PR and solves the same problem:

https://github.com/konstmish/prodigy/pull/19

**Comment by sangoi-exe** [comment]
> Looks absolutely GREAT! What a great idea!
> 
> For anyone curious, this is a more generalized replacement for a previous PR and solves the same problem as [the old one](https://github.com/konstmish/prodigy/pull/19).

Regarding this old PR, it only works with SD15 (as far as I'm aware) since the TEs and Unet can be distinguished by the tensor shape. In SDXL, all the tensors share the same shape, so sadly it won’t work. 🫤

**Comment by Arcitec** [comment]
> In SDXL, all the tensors share the same shape, so sadly it won’t work.

Really? Well, at least FLUX and SD3.5 and stuff are taking over now, and I guess those use different tensor shapes.

**Comment by adefazio** [comment]
I think this needs to be a separate optimizer class, perhaps ProdigyGroupwise. We can't change the main prodigy function in this way as it will break any code bases already using prodigy that use separate parameter groups for other reasons. The theory of prodigy doesn't actually work when using separate learning rates for separate parameter groups, and there are cases I've found in the past doing block-wise prodigy (or other split versions) doesn't work well, so it can't be a drop-in improvement.

Given the high potential impact of this version, I'm supportive of including it in the main prodigy repo as a separate function.

**Comment by sangoi-exe** [comment]
> I think this needs to be a separate optimizer class, perhaps ProdigyGroupwise. We can't change the main prodigy function in this way as it will break any code bases already using prodigy that use separate parameter groups for other reasons. The theory of prodigy doesn't actually work when using separate learning rates for separate parameter groups, and there are cases I've found in the past doing block-wise prodigy (or other split versions) doesn't work well, so it can't be a drop-in improvement.
> 
> Given the high potential impact of this version, I'm supportive of including it in the main prodigy repo as a separate function.

Hey, thanks so much for the response!

I’m happy with whatever you choose, I’d be really glad if my suggestion ends up in the official code somehow. 😊
